### ISSUE TITLE
[Tracking] Import phonebook contacts into New Expensify

### ISSUE THREAD

**Issue by roryabraham** [Expensify/App#47938]
_**Proposal:** Import phonebook contacts into New Expensify_

**Strategy**
The most important metric we want to see grow in the coming months is new user acquisition :chart_with_upwards_trend:.  To facilitate that we need to invest in features that streamline viral user growth and make it easier to pull new users into the app. :busts_in_silhouette:

**Problem**
If you want to use the Expensify Classic app to invite your team or your boss to Expensify, it’s very easy. You sign up, scan a receipt, and an inbox task will appear asking who you want to send it to. If you click on that, it will open your phone’s contact book and allow you to select the contact(s) that you want to invite to Expensify. You can invite people you know in just a few taps, without leaving the app.

There are a number of other places throughout Expensify Classic where you can easily invite people to Expensify, such as when submitting reports.

Meanwhile, in New Expensify, the first few steps of inviting new users are arguably more streamlined. You immediately scan a receipt, and on the very next page you’ll be asked who you want to share it with. However, at that stage, you need to leave the app to find their email or phone number from some other source and manually type or copy/paste it into NewDot. This introduces friction in a critical viral growth moment.

**Solution**
Let’s close the gap on one of the final pieces of feature parity between Expensify Classic (mobile) and NewDot, by importing contacts from your phonebook :closed_book: into New Expensify. This will streamline the process of inviting people you already know into Expensify. Let’s show these imported contacts in any page where we invite new users (start chat, workspace invite, request money participants page, etc…).

All these pages are already set up to invite new users to Expensify via their email or phone number, so the only requirement is to import that information from the contacts app (after obtaining the user’s permission to do so). Because of that, _this project is completely front-end and self-contained_. That means that after we get through an initial high-level design phase, it can be implemented entirely by external contributors.

For UI inspiration, WhatsApp 📞 shows imported contacts in an “Invite to WhatsApp” section.

<details><summary>Issue Owner</summary>Current Issue Owner: @blimpich</details>

**Comment by roryabraham** [comment]
First predesign questions:

- How should we show these?
    - I suggest we just show a new section in selection lists titled `Invite to Expensify`, styled the same as the existing `Recents` and `Contacts` sections:
    
    <img width="377" alt="image" src="https://github.com/user-attachments/assets/99d35b67-36bb-4097-906c-049fe6d6cd67">

- Should we try to support web?
    - I suggest no (for the V1, anyways). There's an [experimental ContactsManager web API](https://developer.mozilla.org/en-US/docs/Web/API/Contact_Picker_API), but it only works on Android atm. Furthermore, it doesn't support importing all contacts and displaying them using our own UI. Instead, it's more like what we have in OldApp that pulls up the native contacts app and allows you to select contacts to import.
    
- Should we try to support desktop?
    - I suggest yes. It may be a bit involved to set up, but fortunately our needs should be pretty simple. It might look something like this:
        - Use the same Swift code to access contacts on macOS and iOS. `CNContactsStore` works on both.
        - Use the same C+code on iOS and macOS to create C++ bindings for the swift code
        - (here's where things start to differ) on desktop, create a Node.js addon with JS bindings to the C++/Swift code
        - Require the Node.js addon in `desktop/main.ts`
        - Use the Electron context bridge to provide a renderer process interface for the Node.js code that can be used to access contacts
        
- Should we use a library for this?
    - There are a couple of options, such as [react-native-contacts](https://github.com/morenoh149/react-native-contacts) and [expo-contacts](https://docs.expo.dev/versions/latest/sdk/contacts/). However, because the native code should be quite simple and none of the existing solutions support (or would be expected to support) Electron, I suggest we just create our own small library for this.

**Comment by roryabraham** [comment]
posted first predesign (saving library/implementation details for a 2nd predesign):

https://expensify.slack.com/archives/C01GTK53T8Q/p1724440169163019

**Comment by trjExpensify** [comment]
Posted [here](https://expensify.slack.com/archives/C06ML6X0W9L/p1725268956185409?thread_ts=1724433121.825939&cid=C06ML6X0W9L), but I don't think this is a project for #wave-collect.

**Comment by roryabraham** [comment]
Moved to #f1-25june2025

**Comment by roryabraham** [comment]
Predesigns two and three:

- https://expensify.slack.com/archives/C01GTK53T8Q/p1724788943149429
- https://expensify.slack.com/archives/C01GTK53T8Q/p1725403445145239

**Comment by roryabraham** [comment]
Summarizing all of our predesign decisions and getting this ready for handoff:

- When revealing users imported from contacts which aren't yet an Expensify contact, we'll show them just like any other user. We won't draw any additional attention to the fact that they don't yet have an Expensify account, and won't introduce a new `Invite` section to selection lists
- Every time `SelectionList` mounts, we'll:
    - verify we have permission to access contacts.
        - If the user has given permission to access contacts, proceed to the next step.
        - unless the user has explicitly denied permission, request permission to access contacts, then proceed to the next step
    - import all phonebook contacts, include them in the `SelectionList` result set
- If a user has explicitly denied permission to access contacts, and have entered a search query which yields no results, then instead of simply showing `No results found`, we would show:
`Not seeing who you're looking for? Want to [import your phonebook contacts](link)?`
- Only iOS and Android are in-scope for the V1. I think the V2 can target desktop as well, _maybe_ web in a (distant) V3 if the web API improves.
- We should show a user's contact photo when searching for them and they don't yet an Expensify account.
- If they _do_ have an Expensify account, but their Expensify account is using a default avatar, _and_ your phonebook contact has an avatar for them, then we should show your phonebook avatar for the user rather than their default Expensify avatar
- Let's also roll-our-own small amount of native code for this directly in E/App rather than using or publishing a separate package. Let's use Nitro Modules for this native code.

**Comment by roryabraham** [comment]
last slack update: https://expensify.slack.com/archives/C035J5C9FAP/p1726770468190059?thread_ts=1725380343.339199&cid=C035J5C9FAP

**Comment by perunt** [comment]
👋🏻

**Comment by s77rt** [comment]
You can use https://github.com/s77rt/react-native-contacts (a Turbo Module) for reference

**Comment by blimpich** [comment]
@perunt how is the PR going? Can you you give us an update?

**Comment by perunt** [comment]
hey @blimpich, we just finished fixing the Nitro issue (the library we use for this module). Tomorrow, I'm going to show some metrics about it and move that module to the Expensify repo

**Comment by blimpich** [comment]
Sounds great! Thank you for the update 👍

**Comment by perunt** [comment]
> You can use https://github.com/s77rt/react-native-contacts (a Turbo Module) for reference

Thanks for the link! I'll use it for an approximate comparison between NitroModule and TurboModule

**Comment by perunt** [comment]
While this isn't the perfect comparison since both modules are built differently under the hood and could probably be optimized further, it gives us a decent starting point.
I ran some tests importing around 1,100 contacts - did it 20 times with 10-second breaks between runs to get the average times. Maybe we don't have huge differences, but we definitely see the cost of overhead. 

|  iPhone XR |  Samsung S21 |
|--------|---------|
| <img src="https://github.com/user-attachments/assets/aa86afe1-6916-4668-b7e2-ff3d7d9d5d3a" height="800"/> | <img src="https://github.com/user-attachments/assets/1d3362c9-9bfd-45f4-94f7-1404df1af54c" height="800"/> |

Since we're getting responses under 100ms, it's actually faster than receiving data from the internet (at least for importing 1k contacts) - it allows us to work without using fetch from the backend (in case when we save it).

**Comment by mrousavy** [comment]
woohooooo!!! huge 🔥

**Comment by shubhamguptadream11** [comment]
@perunt Can you share the nitro module repo link where you did the benchmarks.

**Comment by perunt** [comment]
@shubhamguptadream11 this was my sandbox, [link](https://github.com/perunt/react-native-nitro-contact)
I will update and clean it later, and add screenshots to that repo

**Comment by perunt** [comment]
I'm adding the last features - 'use photo from contact book for contacts from Expensify that don't have an avatar.' The second ability is that when you pick a user from contacts, you create this user for the Expensify account. Due to our current structure, it's a bit tricky to use the photo for creating this contact on the next screen. I mean to port this photo. Apart from this, everything is working.
So I'm starting to think if we need to deliver it in one PR or maybe spread it. I know that if you check the amount of code, it would be around 50k lines, but I want to say that 95% is auto-generated code. I'll also mention this in the PR for reviewer.
What do you think? @blimpich (I mean deliver it as one PR or spread it)

**Comment by blimpich** [comment]
@perunt I think splitting it up is a good idea. I think if that functionality is particularly tricky, it's preferable to break it up so that we can review and test separately. We would just use one of the randomly picked default avatars instead of the contact photo in the first PR correct?

**Comment by perunt** [comment]
@blimpich, it's good for review. We may need to change the view of two buttons. I'll ping the design team

**Comment by DylanDylann** [comment]
@perunt @blimpich Do we support this feature on mWeb?

**Comment by perunt** [comment]
No, it’s only for native mobile. I also left some hints to make the review process easier:

<img width="764" alt="image" src="https://github.com/user-attachments/assets/fbe8a442-ccf9-41ce-bc86-d564ed120a7b">

As we can see, the design will be updated as the design team makes some adjustments. So, I’m marking this as WIP for now

**Comment by shawnborton** [comment]
cc @Expensify/design let's align over here on the design of this. 

One thing we need to consider is that for brand new users, they will likely see something like this as we are going to start showing a test account to submit to, as well as the "Just track it for now" assuming we combine Track & Submit into just "Create expense":

![CleanShot 2024-11-13 at 18 54 57@2x](https://github.com/user-attachments/assets/e45cf3c7-7c5d-4ff1-b409-6559684b3b04)

@dubielzyk-expensify I think you might have been hinting at just proactively firing off the permission dialog to import contacts as soon as you get on this page? 

Otherwise we could do something where we use an option row?

![CleanShot 2024-11-13 at 18 55 38@2x](https://github.com/user-attachments/assets/bbf6bbcd-9efd-4004-bd4b-9ce27a5c5f68)

**Comment by dannymcclain** [comment]
Oh I love your option row for import contacts. That feels right to me.

>>Additionally, has it been decided that we're taking a passive approach to asking for permission instead of showing a dialog?

>What do you mean by that exactly Jon?

In this comment, I think what Jon was referencing was our approach to locations permissions, where we do a "soft ask" (i.e. a branded pop up that tells users we're about to ask for the permission) before straight up throwing the native device permissions dialog. @dubielzyk-expensify can correct me if I'm wrong though!

Here's the idea from locations if that is indeed what he was talking about:
![CleanShot 2024-11-13 at 13 55 58@2x](https://github.com/user-attachments/assets/600fc78a-aa6a-47cf-ae53-91fc1e2477e0)

So how that might apply here (just one idea):
- User clicks on `Import contacts` option row
- RHP opens with some content explaining the benefits of syncing contacts from your device with a big green "LET'S DO IT" button
- User taps "LET'S DO IT" and then we surface the native permissions prompt (like what's being shown in the right screen above, except it would be for contacts obviously)
- They grant the permissions
- We could then either just close the RHP back to the "Submit it to someone" screen or show a little confirmation screen in the RHP and list of the contacts we imported or just show one of our general "Success" screens.

**Comment by blimpich** [comment]
@dannymcclain Should we skip the first step of having them click "import contacts?" Seems like if they click that they are also going to click "LET'S DO IT." Could reduce the number of clicks to reach the same conclusion and just show the RHP when we would otherwise show the little "import contacts" option row?

**Comment by dannymcclain** [comment]
Oh yeah that's a good point. I guess if they click "Import contacts" we could probably skip the soft ask. Or if we don't want to show that option row, we could automatically pop open the soft ask. But I think you're right, with either approach we could probably skip a step.

**Comment by dubielzyk-expensify** [comment]
@shawnborton wouldn't the test contact only open if they tap "Yes" on this test popover? 
![image](https://github.com/user-attachments/assets/344e7246-ac27-4ab8-81f0-4dcd22f60cf9)

If so, I wonder if we should just leave the contacts permission out of the test flow altogether? 

My initial thinking was to auto-pop open the soft ask without the option row of saying "Import contacts":

So the flow would be:
1. User opens to submit a report
2. If test case, do nothing.
3. If real case and they proceed to contacts screen, auto-pop open the soft ask
4. If user says yes, great. We have permission.
5. If user says no, we add the option row in there.

**Comment by shawnborton** [comment]
> wouldn't the test contact only open if they tap "Yes" on this test popover?

My understanding is that we are ALWAYS going to show it no matter what. The popover you are showing is only to upload the fake receipt. So it's possible that you could submit anything (manual, distance) to the fake manager for testing purposes.

**Comment by dubielzyk-expensify** [comment]
Ohhhhhh, interesting. I see. I guess I'd still say the flow above could work. If not, then yeah, option row + hard ask feels right

**Comment by perunt** [comment]
> Oh yeah that's a good point. I guess if they click "Import contacts" we could probably skip the soft ask.

Do we have a place where we’ve skipped the native permission dialog? As far as I know, it's forbidden, and the user needs to interact with it.

The reason I asked this at the beginning is that if the user scrolls a bit and then adds contacts, the whole order of the list will change, which might look flaky. That’s why I’m doing it at the start.
I guess we should mix the order of contacts only after pressing the button or at the very beginning when we were navigated to the screen.

**Comment by shawnborton** [comment]
> Do we have a place where we’ve skipped the native permission dialog? As far as I know, it's forbidden, and the user needs to interact with it.

No plans to skip it to my knowledge, we are aware that we need it. 

> The reason I asked this at the beginning is that if the user scrolls a bit and then adds contacts, the whole order of the list will change, which might look flaky. That’s why I’m doing it at the start.

I don't feel strongly here. I think it's an assumed thing that if you add more rows to the list, the order will change. So I'm personally not super worried here.

**Comment by shawnborton** [comment]
Based on some feedback from the @Expensify/design team, here are some options:

![CleanShot 2024-11-14 at 12 10 14@2x](https://github.com/user-attachments/assets/7c0c61ea-ab92-44ef-aea9-dd218fc4cb00)

![CleanShot 2024-11-14 at 12 10 25@2x](https://github.com/user-attachments/assets/83d27a48-b0dd-4abe-aa4c-f9a82ba3db31)

![CleanShot 2024-11-14 at 12 10 33@2x](https://github.com/user-attachments/assets/9e02c10f-5ca1-4ecd-acea-5b3ff85a4758)

Any preference there? I kinda feel like adding contacts is so straight forward that maybe we don't need the soft ask pattern, though it does look nice.

**Comment by perunt** [comment]
> I don't feel strongly here. I think it's an assumed thing that if you add more rows to the list, the order will change. So I'm personally not super worried here.

All I meant is to avoid doing this randomly so the user is aware it's caused by their action.

Providing contacts can be considered a big step for some users. Maybe we need to emphasize that we don’t save them on the server automatically and retrieve them each time the screen is opened? I really like how the soft ask looks, though.

**Comment by dannymcclain** [comment]
@shawnborton honestly I prefer the immediate soft ask or the soft ask after the option row. I know some people are going to say that soft ask is just an extra step coming from the option row, but for some reason that little moment before the native permissions dialog (even if I've explicitly stated I want to import my contacts via an option row) feels cozy and comforting.

I could also see a flow like Jon described where we pop the soft ask immediately, and if the user chooses `Not now`, we add the Import contacts option row to the list. And then when they tap that, we could either surface the soft ask again (my preference) or just surface the native permissions dialog.

**Comment by dubielzyk-expensify** [comment]
Great stuff @shawnborton ! I really like the immediate soft ask. 

> I could also see a flow like Jon described where we pop the soft ask immediately, and if the user chooses Not now, we add the Import contacts option row to the list. And then when they tap that, we could either surface the soft ask again (my preference) or just surface the native permissions dialog.

Yeah, I think it's slightly weird to see "Import contacts" under the scrim but not a deal breaker at all. Maybe it's weirder for it to appear if they say "No thanks". Dig this direction either way. Feels like it'll convert more than the others.

**Comment by dubielzyk-expensify** [comment]
Just also a note that I wonder if we should be more "value prop-y" on the copy instead of "Import contacts". Maybe something like "Pay your friends or collagues", "Send expenses to your contacts", "Select from contact book" or something. I don't mind "Import contacts" either, but wondering if users are a bit more cautious about it. I'm sure @jamesdeanexpensify would have better ideas though

**Comment by shawnborton** [comment]
Yeah, totally agree we should give the copy a second look. 

> Yeah, I think it's slightly weird to see "Import contacts" under the scrim but not a deal breaker at all.

I think this is probably our best option though, otherwise re-asking the user each time seems pretty invasive.

**Comment by dannymcclain** [comment]
Totally agree, let's give the copy some zhuzh.

>>Yeah, I think it's slightly weird to see "Import contacts" under the scrim but not a deal breaker at all.

>I think this is probably our best option though, otherwise re-asking the user each time seems pretty invasive.

I don't think it's _that_ weird, and agree that this is probably our best option unless we want to move that option row somewhere into settings 🙈.

So let me see if I've got this summary right:

- Add an `Import contacts` option row to the contact list
- Open the "soft ask" modal immediately when someone gets to the Choose a contact screen
- If they tap `Continue`, we surface the native permissions prompt
- If they tap `Not now`, we dismiss the soft ask and do nothing

Does that sound right?

Follow up questions:
- If the user grants us contact permissions, do we hide the `Import contacts` option row?
- What happens when the user taps the `Import contacts` option row?
  - Do we open the soft ask again or do we go straight to native permissions?

**Comment by perunt** [comment]
One more question: should we use the same row/button for importing contacts on the empty search page as we do in the main list?

Since the empty screen provides a bit more space, we could also include an explanation of why the user might need this feature. For example, as mentioned earlier: `Tap to import your contacts and quickly choose someone to share this expense with!`
It would also be nice to get a Spanish translation for all the text

**Comment by blimpich** [comment]
Ah good reminder, I will make sure to get you the Spanish copy once we have finalized the design 👍

**Comment by shawnborton** [comment]
All of this discussion sounds great and I am aligned with where your conclusions @dannymcclain 

In terms of showing the import contacts row on an empty search... part of me thinks we shouldn't do that just yet, so we aren't competing with this empty state that we recently built:
![CleanShot 2024-11-15 at 19 05 42@2x](https://github.com/user-attachments/assets/0ab0a525-fb94-4aaa-bdb9-0f990cdbe61d)

**Comment by jamesdeanexpensify** [comment]
(let me know when/if you need any copy input, saw my tag a while back!)

**Comment by blimpich** [comment]
@jamesdeanexpensify yes! Could you look at the [mockups here](https://github.com/Expensify/App/issues/47938#issuecomment-2476065217) and give your opinion/feedback on the copy? Anything you'd change?

**Comment by jamesdeanexpensify** [comment]
Happy to leave a bit of feedback! One question - is all this fair game, or only the body text? 

<img width="122" alt="image" src="https://github.com/user-attachments/assets/615808c0-7745-4225-8b15-8278bd39aba9">

**Comment by dannymcclain** [comment]
@jamesdeanexpensify that one is the native device prompt and AFAIK we cannot change anything about it.

**Comment by blimpich** [comment]
I think all is fair game with the exception of `"New Expensify" Would like to Access Your Contacts`

**Comment by shawnborton** [comment]
So something like this? 
![CleanShot 2024-11-18 at 08 36 44@2x](https://github.com/user-attachments/assets/8e8a9c61-9544-4095-8e9f-60cfc8505199)

**Comment by perunt** [comment]
> In terms of showing the import contacts row on an empty search... part of me thinks we shouldn't do that just yet, so we aren't competing with this empty state that we recently built:

We could still keep it very minimalistic—perhaps just extending 'No results found' with 'Import contacts?'.

The reason I'm so focused on this feature is that when a user has 200-300 contacts in their list, the only way to see the import button would be by scrolling all the way to the end. I feel this approach might rarely be seen by users.
But if you’d prefer to skip this step, we can do that.

> So something like this?

I like it. I can implement the soft input right away since we agreed that it’s a nice addition here. I’ll wait for approvals on the rest.

**Comment by perunt** [comment]
Somehow, I can't get the location soft permission. Do we have any other place in the app where soft permissions are implemented?

**Comment by dubielzyk-expensify** [comment]
> Somehow, I can't get the location soft permission. Do we have any other place in the app where soft permissions are implemented?

Not that I'm aware of. We do technically do it when you scan without camera permission, but it doesn't use the same UI/bottom sheet.

**Comment by perunt** [comment]
I just realized something: since we have a keyboard there, we need to adjust the input focus behavior.

If we need to show the soft permission prompt, we should block input focusing and prevent the keyboard from appearing. Once the user selects 'Allow' or 'Deny,' we can proceed with focusing the input and showing the keyboard.

Does that sound reasonable to you?
![Знімок екрана 2024-11-18 о 11 42 00](https://github.com/user-attachments/assets/8d47b46c-9487-414b-8487-6202996b50d1)

**Comment by dannymcclain** [comment]
>We could still keep it very minimalistic—perhaps just extending 'No results found' with 'Import contacts?'.

I'm honestly not against something like this. It might be a tidy way to tuck this in. Something like:
![CleanShot 2024-11-18 at 08 26 26@2x](https://github.com/user-attachments/assets/77a79665-dbd7-4120-bece-26c788e91388)
(We can of course make the copy more concise with some help. It's feeling a _little busy_ with both pieces of text including a link. But I don't think the idea is a hard no from me.)

>I just realized something: since we have a keyboard there, we need to adjust the input focus behavior.
>If we need to show the soft permission prompt, we should block input focusing and prevent the keyboard from appearing. Once the user selects 'Allow' or 'Deny,' we can proceed with focusing the input and showing the keyboard.
>Does that sound reasonable to you?

This does sound reasonable to me.

**Comment by perunt** [comment]
> I'm honestly not against something like this. It might be a tidy way to tuck this in. Something like:

I'm down for something like this - nice way to fit it in. Love that it keeps users informed but doesn't mess with how the page looks.

**Comment by blimpich** [comment]
Added @DylanDylann since they're the C+ on the draft PR

**Comment by dubielzyk-expensify** [comment]
> I'm honestly not against something like this. It might be a tidy way to tuck this in. Something like:

Tidy mock! Dig it

> I just realized something: since we have a keyboard there, we need to adjust the input focus behavior.
If we need to show the soft permission prompt, we should block input focusing and prevent the keyboard from appearing. Once the user selects 'Allow' or 'Deny,' we can proceed with focusing the input and showing the keyboard.
Does that sound reasonable to you?

Yep, I'm with Danny here. That sounds resonable.

**Comment by shawnborton** [comment]
Agree about the keyboard behavior - makes sense and works for me. And I like Danny's mock - I think that's a nice way to promote it without needing to adjust that empty state too much.

**Comment by blimpich** [comment]
@perunt here are the approved Spanish translations for the main copy. Can you update the PR with these? I can see from the `en.ts` file there are a couple more translations needed than this. I will get those to you soon.

**English:** Import your contacts
**Spanish:** Importa tus contactos

**English:** Import contacts from your phone so your favorite people are always a tap away.
**Spanish:** Importa contactos desde tu teléfono para que tus personas favoritas siempre estén a un toque de distancia.

**English:** Just one more step! Give us the green light to import your contacts.
**Spanish:** ¡Solo un paso más! Danos luz verde para importar tus contactos.

**Comment by blimpich** [comment]
@perunt where do we plan to use [this copy](https://github.com/Expensify/App/pull/51018/files#diff-2a9066512656fa29b1565e362127de5cc11ba95b28f143a9db26f76dc3750785R1053)?

**Comment by perunt** [comment]
I just realized that the app lacks translations for native alerts, such as location and camera, etc. In all these cases, English localization is being used. I wrote about it in slack ([link](https://margelo.slack.com/archives/C035J5C9FAP/p1732104453050019))

**Comment by perunt** [comment]
> @perunt I think we're not using [this copy](https://github.com/Expensify/App/pull/51018/files#diff-2a9066512656fa29b1565e362127de5cc11ba95b28f143a9db26f76dc3750785R1053) anymore? Do we need approved translations for this?

You provided me with an almost completed version of the translation. The last one I used is:
```
English: Import contacts
Spanish: Importa contactos
```
If you can approve this translation, then I have everything I need.

**Comment by blimpich** [comment]
@perunt small change, need an "r" at the end of "Importa":
```
English: Import contacts
Spanish: Importar contactos
```

**Comment by perunt** [comment]
@blimpich I wouldn’t pause on this. Currently, it only creates a user with the email. Once you update the backend, I’ll just need to pass the proper value, and that’s it. So, I guess it would be nice to have this PR 'ready to merge,' then update just 3-4 lines of code and get it merged after a small recap

**Comment by blimpich** [comment]
@perunt agreed, lets keep moving it forward then 👍

**Comment by blimpich** [comment]
@DylanDylann are you able to re-review the PR soon?

**Comment by DylanDylann** [comment]
Sure, but the PR is big, let give me time to thoroughly review it

**Comment by DylanDylann** [comment]
> The initial idea by @roryabraham was to keep it within the project. Probably 90-95% of the entire code is generated, and almost everything could be reused for the next modules. I would rather see this as two native files and two specs for them. This way, there are only a few changes we need to maintain. However, I understand your concerns.

@perunt Thanks for your explanation. I noticed this is the first time we do that in the App repository. @blimpich Just want to confirm again 

cc @roryabraham

**Comment by blimpich** [comment]
We need a bug zero member on this issue.

**Comment by blimpich** [comment]
@zanyrenney we should pay @DylanDylann $250 for reviewing https://github.com/Expensify/App/pull/51018. It was reverted but we had to revert for reasons unrelated to the changes in the PR, so there should be no penalty.

**Comment by blimpich** [comment]
**Next steps:**
- [ ] @perunt will submit the PR again
- [ ] we'll merge the PR once this blocking issue is closed: https://github.com/Expensify/App/issues/49845
- [ ] we'll create a follow up PR to fix the problem talked about in [this comment](https://github.com/Expensify/App/pull/51018#issuecomment-2527589271)

**Comment by blimpich** [comment]
@Julesssss looks like the PRs for https://github.com/Expensify/App/issues/49845 are deployed and in their regression testing period. Should we wait till the end of that before merging in [this big PR that we reverted](https://github.com/Expensify/App/pull/54209)? Or do you think we're good to merge in now?

**Comment by Julesssss** [comment]
I think you're good to go now. Please make sure this is tested locally against HybridApp builds though, as any new issues with the build process could block App development just as we are OOO for the holidays.

**Comment by perunt** [comment]
I've tested it on my end. @DylanDylann also finished his testing. So I think we're good to go, @blimpich.

**Comment by blimpich** [comment]
I have to wait to merge unfortunately. We're having a fire right now related to our hybrid app and merging a big PR is likely going to get reverted again. I'll try to get it in this week and confirm that it won't get reverted again, but probably won't happen today.

**Comment by perunt** [comment]
By the way, I’d be happy to help with this as well. We have some extra cycles at Margelo, so feel free to reach out if you need anything!

**Comment by blimpich** [comment]
Merged again. Hoping this time nothing happens to get it reverted 🤞🍀

**Comment by Julesssss** [comment]
> Merged again. Hoping this time nothing happens to get it reverted 🤞🍀

Hey all, I just wanted to check this was tested on HybridApp iOS and Android?

**Comment by DylanDylann** [comment]
As C+ reviewer, I only test on Newdot

**Comment by perunt** [comment]
Hey @Julesssss
I don’t have access to the hybrid app repo yet. Maybe I can run it through a GitHub Action and test it out?

**Comment by zanyrenney** [comment]
Hey Ben, can you let me know what the updated payment date should be on this if it was reverted? @blimpich

**Comment by blimpich** [comment]
> Hey all, I just wanted to check this was tested on HybridApp iOS and Android?

It was not, sorry @Julesssss I completely blanked on checking that 😞. You very clearly asked to have us check that, that is 100% my bad.

> Hey Ben, can you let me know what the updated payment date should be on this if it was reverted?

@zanyrenney if this doesn't get reverted again (it might), then December 25th would be the pay out date I believe.

**Comment by WoLewicki** [comment]
> It was not, sorry @Julesssss I completely blanked on checking that 😞. You very clearly asked to have us check that, that is 100% my bad.

Yeah the problem is that it's not compatible with Hybrid App right now and it might not be too easy to fix since we don't have much knowledge of the internals of Nitro Modules. We'll try to make it work though 🚀

**Comment by Julesssss** [comment]
Yeah it's pretty difficult for anyone to test right now. Especially just after the submodule change and other build issues 😭

**Comment by blimpich** [comment]
@zanyrenney we ended up reverting again fyi.

@perunt @DylanDylann work is being done to make it so that you can test Hybrid App in the near future, but lots of Expensify Engineers are taking time off this time of year so we might have to wait till January. We're close to getting working ad-hoc builds [here](https://github.com/Expensify/App/issues/53843), so hopefully that can be used soon.

Will keep you updated on adding you to the Mobile-Expensify repo. Takes time. We're revamping some internal processes to make this easier in the future. Sorry about all the reverts 😢 

@perunt no rush, but when you get a chance can you re-raise the PR?

In the meantime, I'm going to change this to weekly cadence.

**Comment by perunt** [comment]
Okay, I'm going to open it again. It's funny how many times it has been reverted, especially since there's no direct issue with the PR 🥲 
This time, we’ll definitely get it merged 😸

**Comment by perunt** [comment]
Okay, I'm going to open it again. It's funny how many times it has been reverted, especially since there's no direct issue with the PR 🥲
This time, we’ll definitely get it merged 😸

> Will keep you updated on adding you to the Mobile-Expensify repo. Takes time. We're revamping some internal processes to make this easier in the future. Sorry about all the reverts 😢

It will make things easier, thank you!

**Comment by blimpich** [comment]
## Update:

After being reverted we have [a PR up](https://github.com/Expensify/App/pull/54459) but we're waiting for @perunt to be added to the Mobile Expensify Repo so that we can thoroughly test the PR to ensure it doesn't break HybridApp. There is an open issue tracking adding new devs to the Mobile-Expensify repo [here](https://github.com/Expensify/App/issues/54321). Progress has been slow since lots of people are out for holidays.

**Comment by perunt** [comment]
@blimpich the Contact Module has been successfully reapplied. Additionally I've bumped Nitro.
Since I still don't have access to the Hybrid App, could you please assign someone who can run it? Thanks!

**Comment by blimpich** [comment]
Reached out to engineers familiar with hybrid app here: https://expensify.slack.com/archives/C04878MDF34/p1736210982318679.

**Comment by Julesssss** [comment]
Hey @blimpich, @staszekscp submitted a promising-looking PR for resolving issues [here](https://github.com/Expensify/Mobile-Expensify/pull/13347), which was just merged 🎉

**Comment by blimpich** [comment]
@perunt can you look at the most recent comments on the PR and merge in `main`? I think we're about ready to merge again.

**Comment by blimpich** [comment]
@DylanDylann are you able to test the PR for hybrid app compatibility? https://github.com/Expensify/App/pull/54459

**Comment by DylanDylann** [comment]
@blimpich I am trying to setup a hybrid app locally. Could we build ad-hoc for hybrid App?

**Comment by blimpich** [comment]
### Update

Everyone on this issue has access to the hybrid-app now. @perunt is working on fixing build issues in [this PR](https://github.com/Expensify/App/pull/54459).

**Comment by perunt** [comment]
All good, Melvin. A few days ago, I checked the PR, and it works for HybridApp, so I requested a review on it: https://github.com/Expensify/App/pull/54459#issuecomment-2674679190

**Comment by blimpich** [comment]
@DylanDylann do you have an ETA on when you'll be able to give the new PR a review?

**Comment by roryabraham** [comment]
rejoining the fray as a reviewer here now that I'm back from parental leave

**Comment by blimpich** [comment]
### Update:

See [PR](https://github.com/Expensify/App/pull/54459#issuecomment-2734590968), @perunt is actively working on this.

**Comment by blimpich** [comment]
@perunt can you provide an update here?

**Comment by blimpich** [comment]
@perunt can you provide an update here?

**Comment by perunt** [comment]
@blimpich just resolved all conflicts altogether with the new ESLint rules and Rory's comments. So now it's good to go

**Comment by blimpich** [comment]
### Update

Progress is being made. [PR](https://github.com/Expensify/App/pull/54459) is approved by C+. We're working on getting QA to do a regression suite on this and then also having robust test steps for it ([link](https://expensify.slack.com/archives/C9YU7BX5M/p1747260442439719)).

**Comment by perunt** [comment]
QA steps are there
I'm going to resolve these bug reports

**Comment by perunt** [comment]
I reviewed bugs and almost all of them came from the main branch and disappeared after the merge
The only thing that is left is this stuff ([1 and 11 bug](https://github.com/Expensify/App/pull/54459#issuecomment-2894500824)) that I want to hear your opinion on @roryabraham

**Comment by roryabraham** [comment]
Now that the first usage of this is on prod and has passed regression tests, I'm creating a couple PRs to remove the beta in the front-end and the back-end

**Comment by roryabraham** [comment]
We also added only one usage of contact import, and we need to add it in more places.

**Comment by mallenexpensify** [comment]
@roryabraham , saw "Due for payment 2025-06-18" and that Zany was assigned as BZ.  She's OOO, please tag me if a payment is due this week.   I'm removing the payment pretext in the title too.

**Comment by mallenexpensify** [comment]
Contributor+: @DylanDylann due $1500 via NewDot
